File: maximum-product-subarray/solution152.py
# https://leetcode.com/problems/maximum-product-subarray/

from typing import List


class Solution152:

    # T: O(n), S: O(1)
    def maxProduct(self, nums: List[int]) -> int:
        if not nums:
            return 0

        n = len(nums)
        max_product = nums[0]
        current_product = 1

        for i in range(n):
            current_product *= nums[i]
            max_product = max(max_product, current_product)

            if current_product == 0:
                current_product = 1

        current_product = 1
        for i in range(n - 1, -1, -1):
            current_product *= nums[i]
            max_product = max(max_product, current_product)

            if current_product == 0:
                current_product = 1

        return max_product

    # T: O(n), S: O(1)
    def maxProductKadane(self, nums: List[int]) -> int:
        if not nums:
            return 0

        max_so_far = nums[0]
        min_so_far = nums[0]
        result = max_so_far

        for i in range(1, len(nums)):
            curr = nums[i]
            temp_max = max(curr, max(max_so_far * curr, min_so_far * curr))
            temp_min = min(curr, min(max_so_far * curr, min_so_far * curr))

            max_so_far = temp_max
            min_so_far = temp_min

            result = max(result, max_so_far)

        return result

    # T: O(n^2), S: O(1)
    def maxProductBrute(self, nums: List[int]) -> int:
        if not nums:
            return 0

        n = len(nums)
        result = nums[0]
        for i in range(n):
            cur = nums[i]
            result = max(result, cur)
            for j in range(i + 1, n):
                cur *= nums[j]
                result = max(result, cur)
        return result


def main() -> None:
    sol = Solution152()
    nums1 = [2, 3, -2, 4]
    nums2 = [-2, 0, -1]
    print(sol.maxProduct(nums1))       # 6
    print(sol.maxProductKadane(nums1))  # 6
    print(sol.maxProductBrute(nums1))  # 6
    print(sol.maxProduct(nums2))       # 0
    print(sol.maxProductKadane(nums2))  # 0
    print(sol.maxProductBrute(nums2))  # 0


if __name__ == "__main__":
    main()

File: maximum-product-subarray/solution152.go
// https://leetcode.com/problems/maximum-product-subarray/

package main

import (
	"fmt"
)

// T: O(n), S: O(1)
func maxProduct(nums []int) int {
	if len(nums) == 0 {
		return 0
	}

	n := len(nums)
	maxProduct := nums[0]
	currentProduct := 1

	for i := 0; i < n; i++ {
		currentProduct *= nums[i]
		maxProduct = max(maxProduct, currentProduct)

		if currentProduct == 0 {
			currentProduct = 1
		}
	}

	currentProduct = 1
	for i := n - 1; i >= 0; i-- {
		currentProduct *= nums[i]
		maxProduct = max(maxProduct, currentProduct)

		if currentProduct == 0 {
			currentProduct = 1
		}
	}

	return maxProduct
}

// T: O(n), S: O(1)
func maxProductKadane(nums []int) int {
	if len(nums) == 0 {
		return 0
	}

	maxSoFar := nums[0]
	minSoFar := nums[0]
	result := maxSoFar

	for i := 1; i < len(nums); i++ {
		curr := nums[i]
		tempMax := max(curr, max(maxSoFar*curr, minSoFar*curr))
		tempMin := min(curr, min(maxSoFar*curr, minSoFar*curr))

		maxSoFar = tempMax
		minSoFar = tempMin

		result = max(result, maxSoFar)
	}

	return result
}

// T: O(n^2), S: O(1)
func maxProductBrute(nums []int) int {
	if len(nums) == 0 {
		return 0
	}

	n := len(nums)
	result := nums[0]
	for i := 0; i < n; i++ {
		cur := nums[i]
		result = max(result, cur)
		for j := i + 1; j < n; j++ {
			cur *= nums[j]
			result = max(result, cur)
		}
	}
	return result
}

func main() {
	nums1 := []int{2, 3, -2, 4}
	nums2 := []int{-2, 0, -1}
	fmt.Println(maxProduct(nums1))       // 6
	fmt.Println(maxProductKadane(nums1)) // 6
	fmt.Println(maxProductBrute(nums1))  // 6
	fmt.Println(maxProduct(nums2))       // 0
	fmt.Println(maxProductKadane(nums2)) // 0
	fmt.Println(maxProductBrute(nums2))  // 0
}
